parser/ParseBlockProperties.hpp

```diff
@@ -0,0 +1,254 @@
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_PARSER_PARSE_BLOCK_PROPERTIES_HPP_
+#define QUICKSTEP_PARSER_PARSE_BLOCK_PROPERTIES_HPP_
+
+#include <cstdint>
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "parser/ParseKeyValue.hpp"
+#include "parser/ParseString.hpp"
+#include "parser/ParseTreeNode.hpp"
+#include "utility/Macros.hpp"
+#include "utility/PtrList.hpp"
+#include "utility/StringUtil.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+
+/** \addtogroup Parser
+ *  @{
+ */
+
+/**
+ * @brief Encapsulates the BlockProperties key-value list. Makes the job
+ *        of resolving BlockProperties easy.
+ */
+class ParseBlockProperties : public ParseTreeNode {
+  // Valid key names for the BlockProperties.
+  static const std::string kKeyBlockSizeMB;
+  static const std::string kKeyCompress;
+  static const std::string kKeySort;
+  static const std::string kKeyType;
+
+ public:
+  /**
+   * @brief Constructor.
+   * 
+   * @param line_number Beginning line number.
+   * @param column_number Beginning column number.
+   * @param properties PtrList to the KeyValues.
+   */
+  ParseBlockProperties(int line_number,
+                       int column_number,
+                       PtrList<ParseKeyValue> *properties)
+      : ParseTreeNode(line_number, column_number),
+        properties_(properties) { }
+
+  /**
+   * @return The name of this entity.
+   */
+  std::string getName() const override {
+    return "BlockProperties";
+  }
+
+  /**
+   * @brief Returns the first repeated key contained in the key-value list.
+   * @details A repeated key is any 2 key-values with case-insensitive matching
+   *          key names.
+   *          
+   * @return A pointer to the first repeated key-value or nullptr if there
+   *         are no repetitions.
+   */
+  const ParseKeyValue* getFirstRepeatedKeyValue() const {
+    std::set<std::string> seen_keys;
+    for (const ParseKeyValue &key_value : *properties_) {
+      std::string lower_key = ToLower(key_value.key()->value());
+      std::set<std::string>::iterator itr = seen_keys.find(lower_key);
+      if (itr != seen_keys.end()) {
+        return &key_value;
+      }
+      seen_keys.insert(lower_key);
+    }
+    return nullptr;
+  }
+
+  /**
+   * @brief Returns the first invalid key contained in the key-value list.
+   * @details An invalid key-value is a key-value with a key whose case-insensitive
+   *          string does not match one of the valid key names.
+   *          
```

Comment:
@cramja There are trailing spaces in this line, and other places in this file. Please fix in another PR.